OSDOCS-9812: add custom CA certs MicroShift (#74615)
openshift/openshift-docs: modules/microshift-custom-ca-proc.adoc

// Module included in the following assemblies:
//
// * microshift_security_compliance/microshift-custom-ca.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-custom-cas-configuring_{context}"]
= Configuring custom certificate authorities

To configure externally generated certificates and domain names using custom certificate authorities (CAs), add them to the {microshift-short} `/etc/microshift/config.yaml` configuration file. You must also configure the host operating system trust root.

[NOTE]
====
Externally generated `kubeconfig` files are created in the `/var/lib/microshift/resources/kubeadmin/<hostname>/kubeconfig` directory. If you need to use `localhost` in addition to externally generated configurations, retain the original `kubeconfig` file in its default location. The `localhost` `kubeconfig` file uses the self-signed certificate authority.
====

.Prerequisites
* The OpenShift CLI (`oc`) is installed.
* You have access to the cluster as a user with the cluster administration role.

> * Certificate authorities are set up.

Reviewer: What does this mean? Does it mean that the new CA is added to the trust store of the microshift host? If yes, we should make this more explicit.

* A {microshift-short} `/etc/microshift/config.yaml` configuration file exists.

.Procedure

> . Copy the externally generated certificates you want to add to the {microshift-short} host, ensuring that the files are accessible only by the trust root of the operating system.

Reviewer: "accessible only by the trust root of the os" is confusing.
Actually it is:
- the CA that has issued the custom CA must be added to the trust root of the OS
- the cert and private key must be accessible only by root / microshift.

    server: https://api.ci-ln-k0gim2b-76ef8.aws-2.ci.openshift.org:6443
  name: ci-ln-k0gim2b-76ef8
contexts:
- context:
    cluster: ci-ln-k0gim2b-76ef8
    user:
  name:
current-context:
kind: Config
preferences: {}
----
<1> The `certificate-authority-data` section is not present in externally generated `kubeconfig` files. It is added with the `oc config set` command used previously.

.. Verify the `subject` and `issuer` of your customized API server certificate authority by running the following command:
+
[source,terminal]
----
$ curl --cacert /tmp/caCert.pem https://${fqdn_name}:6443/healthz -v
----
+
.Example output
----
Server certificate:
  subject: CN=kas-test-cert_server
  start date: Mar 12 11:39:46 2024 GMT
  expire date: Mar 12 11:39:46 2025 GMT
  subjectAltName: host "dhcp-1-235-3.arm.eng.rdu2.redhat.com" matched cert's "dhcp-1-235-3.arm.eng.rdu2.redhat.com"
  issuer: CN=kas-test-cert_ca
  SSL certificate verify ok.
----
+
[IMPORTANT]
====
Either replace the `certificate-authority-data` in the generated `kubeconfig` file with the new `rootCA` or add the `certificate-authority-data` to the trust root of the operating system. Do not use both methods.
====

> .. Configure additional CAs in the trust root of the operating system. For example, in the RHEL Client truststore on the client system.

Reviewer: maybe add "see referenced RHEL doc on how to do that".

.Procedure

. Copy the externally generated certificates you want to add to the {microshift-short} host, ensuring that the files are accessible only by the trust root of the operating system.

. For each custom CA that you need, add an `apiServer` section called `namedCertificates` to the `/etc/microshift/config.yaml` {microshift-short} configuration file by using the following example:
+
[source,yaml]
----
apiServer:
  namedCertificates:
   - certPath: ~/certs/api_fqdn_1.crt <1>
     keyPath:  ~/certs/api_fqdn_1.key <2>
   - certPath: ~/certs/api_fqdn_2.crt
     keyPath:  ~/certs/api_fqdn_2.key
     names: <3>
     - api_fqdn_1
     - *.apps.external.com
----
<1> Add the full path to the certificate.
<2> Add the full path to the certificate key.
<3> Optional. Add a list of explicit DNS names. Leading wildcards are allowed. If no names are provided, the implicit names are extracted from the certificates.

. Restart the {microshift-service} to apply the certificates by running the following command:
+
[source,terminal]
----
$ systemctl microshift restart
----

. Wait five minutes for system restart and apply the custom server. New `kubeconfig` files are generated.

> . Copy the `kubeconfig` files to the client. If you specified wildcards for the IP address, update the `kubeconfig` to remove the public IP address of the server and replace that IP address with the specific wildcard range you want to use.

Reviewer 2: we should specify where the file should be located in Microshift (source):
ie: /var/lib/microshift/resources/kubeadmin/<custom_cert_fqdn>/kubeconfig

> . From {microshift-short}, specify the `kubeconfig` to use by running the following command:

Reviewer 2: this should be done from the client not from the Microshift,
better to move this section as sub-bullet before 7.a

+
[source,terminal]
----

> $ export KUBECONFIG=/var/lib/microshift/resources/kubeadmin/<custom_cert_fqdn>/kubeconfig

Reviewer 2: the path depends on  the kubeconfig  destination after  it was copied..

[source,terminal]
----
$ export KUBECONFIG=~/custom-kubeconfigs/kubeconfig <1>
----
<1> Use the location of the copied `kubeconfig` file as the path.

.. Check that the certificates are applied by using the following command:
+
[source,terminal]
----

> $ oc --certificate-authority ~/certs/ca.ca get node

Reviewer 3: @ShaunaDiaz for me this is not working
oc --certificate-authority /tmp/caCert.pem get node

Reviewer 3: Sorry it works fine I misplaced new kubeconfig

----
+
.Example output
[source,terminal]
----
oc get node
NAME                             STATUS   ROLES                         AGE   VERSION
dhcp-1-235-195.arm.example.com   Ready    control-plane,master,worker   76m   v1.29.2
----

.. Add the new CA file to the $KUBECONFIG environment variable by running the following command:
+
[source,terminal]
----

> $ oc config set clusters.microshift.certificate-authority /tmp/certificate-authority-data-new.crt

Reviewer 3: @ShaunaDiaz Better to add how user can generate /tmp/certificate-authority-data-new.crt?

Reviewer 3: Others are lgtm from my side

Reviewer 2: well `/tmp/certificate-authority-data-new.crt` is just our example , users  have their own CA cert files. we do need to specify that this file is CA. 
@gangwgr ,
we cannot extract certificate-authority-data using  `oc config view` , because  the generated kubeconfig doesnt have that value. i dont know how that command work for you .. did you  export the generated  KUBECONFIG ?

Reviewer 3: Thanks @eslutsky Got your point

* Certificate authorities are set up.
* A {microshift-short} `/etc/microshift/config.yaml` configuration file exists.

.Procedure

. Copy the externally generated certificates you want to add to the {microshift-short} host, ensuring that the files are accessible only by the trust root of the operating system.

. For each custom CA that you need, add an `apiServer` section called `namedCertificates` to the `/etc/microshift/config.yaml` {microshift-short} configuration file by using the following example:
+
[source,yaml]
----
apiServer:
  namedCertificates:
   - certPath: ~/certs/api_fqdn_1.crt <1>
     keyPath:  ~/certs/api_fqdn_1.key <2>
   - certPath: ~/certs/api_fqdn_2.crt
     keyPath:  ~/certs/api_fqdn_2.key
     names: <3>
     - api_fqdn_1
     - *.apps.external.com
----
<1> Add the full path to the certificate.
<2> Add the full path to the certificate key.
<3> Optional. Add a list of explicit DNS names. Leading wildcards are allowed. If no names are provided, the implicit names are extracted from the certificates.

. Restart the {microshift-service} to apply the certificates by running the following command:
+
[source,terminal]
----
$ systemctl microshift restart
----

> . Wait five minutes for system restart and apply the custom server. New `kubeconfig` files are generated in the `/var/lib/microshift/resources/kubeadmin/` directory.

Author: it was in the polarion; I assumed it takes time for the server to apply

. Copy the `kubeconfig` files to the client. If you specified wildcards for the IP address, update the `kubeconfig` to remove the public IP address of the server and replace that IP address with the specific wildcard range you want to use.

. From the client, use the following steps:

.. Specify the `kubeconfig` to use by running the following command:
+
[source,terminal]
----

> $ export KUBECONFIG=/var/lib/microshift/resources/kubeadmin/<custom_cert_fqdn>/kubeconfig <1>

Reviewer 2: ```suggestion
# assuming `kubeconfig` were copied to ~/custom-kubeconfigs/kubeconfig
$ export KUBECONFIG=~/custom-kubeconfigs/kubeconfig <1>
```

----
<1> Use the location of the copied `kubeconfig` file as the path.

.. Check that the certificates are applied by using the following command:
+
[source,terminal]
----
$ oc --certificate-authority ~/certs/ca.ca get node
----
+
.Example output
[source,terminal]
----
oc get node
NAME                             STATUS   ROLES                         AGE   VERSION
dhcp-1-235-195.arm.example.com   Ready    control-plane,master,worker   76m   v1.29.2
----

.. Add the new CA file to the $KUBECONFIG environment variable by running the following command:
+
[source,terminal]
----
$ oc config set clusters.microshift.certificate-authority /tmp/certificate-authority-data-new.crt
----

> .. Verify that the new certificate was generated by running the following command:

Reviewer 2: ```suggestion
.. Verify that  kubeconfig contain the new CA by running the following command:
```

+
[source,terminal]
----
$ oc config view --flatten
----
+
.Example externally generated `kubeconfig` file
+
[source,yaml]
----
apiVersion: v1
clusters:

> - cluster:
>     server: https://api.ci-ln-k0gim2b-76ef8.aws-2.ci.openshift.org:6443

Reviewer 2: ```suggestion
- cluster:
    certificate-authority: /tmp/certificate-authority-data-new.crt
    server: https://api.ci-ln-k0gim2b-76ef8.aws-2.ci.openshift.org:6443
```

  name: ci-ln-k0gim2b-76ef8
contexts:
- context:
    cluster: ci-ln-k0gim2b-76ef8
    user:
  name:
current-context:
kind: Config
preferences: {}
----
+
[NOTE]
====

> The `certificate-authority-data` section is not present in externally generated `kubeconfig` files.

Reviewer 2: this warning is no longer valid ,since we added the `certificate-authority-data`  in the previous step, please remove it

Author: it is not in the example file that preceeds the Note

====

.Procedure

. For each custom CA that you need, add an `apiServer` section called `namedCertificates` to the `/etc/microshift/config.yaml` {microshift-short} configuration file by using the following example:
+
[source,yaml]
----
apiServer:
  namedCertificates:
   - certPath: ~/certs/api_fqdn_1.crt <1>
     keyPath:  ~/certs/api_fqdn_1.key <2>
   - certPath: ~/certs/api_fqdn_2.crt
     keyPath:  ~/certs/api_fqdn_2.key
     names: <3>
     - api_fqdn_1
     - *.apps.external.com
----
<1> Add the full path to the certificate.
<2> Add the full path to the certificate key.
<3> Optional. Add a list of explicit DNS names. Leading wildcards are allowed. If no names are provided, the implicit names are extracted from the certificates.

. Restart the {microshift-service} to apply the CAs by running the following command:
+
[source,terminal]
----
$ systemctl microshift restart
----

. Wait five minutes for system restart and apply the custom server.

. Specify the `kubeconfig` to use by running the following command:
+
[source,terminal]
----
$ export KUBECONFIG=/var/lib/microshift/resources/kubeadmin/<custom_cert_fqdn>/kubeconfig
----

> //Q: skip this step also?

Reviewer 2: we should   leave this step, its verifies the generated kubeconfig but it shouldnt be done on microshift host, it should be run from the external client.

so the flow  should be:
1. Microshift generate kubeconfig
2. we copy the generated kubeconfig to the client
3. we test the kubeconfig from the client prospective using `oc` command.

====

.Procedure

. For each custom CA that you need, add an `apiServer` section called `namedCertificates` to the `/etc/microshift/config.yaml` {microshift-short} configuration file by using the following example:
+
[source,yaml]
----
apiServer:
  namedCertificates:
   - certPath: ~/certs/api_fqdn_1.crt <1>
     keyPath:  ~/certs/api_fqdn_1.key <2>
   - certPath: ~/certs/api_fqdn_2.crt
     keyPath:  ~/certs/api_fqdn_2.key
     names: <3>
     - api_fqdn_1
     - *.apps.external.com
----
<1> Add the full path to the certificate.
<2> Add the full path to the certificate key.
<3> Optional. Add a list of explicit DNS names. Leading wildcards are allowed. If no names are provided, the implicit names are extracted from the certificates.

. Restart the {microshift-service} to apply the CAs by running the following command:
+
[source,terminal]
----
$ systemctl microshift restart
----

. Wait five minutes for system restart and apply the custom server.

. Check that the certificates are applied by using the following command:
+
[source,terminal]
----

> $ oc --insecure-skip-tls-verify get node

Reviewer 2: 1. before using the `oc` command we need to specify the kubeconfig `export KUBECONFIG=/var/lib/microshift/resources/kubeadmin/<custom_cert_fqdn>/kubeconfig`
2. instead of using insecure we could explicitly specify the CA.
`oc --certificate-authority ~/certs/ca.ca get node`

Author: Chuckling at a file called `ca.ca`.

----

. Restart the {microshift-service} by running the following command:
+
[source,terminal]
----
$ systemctl microshift restart
----

. Verify that the new certificate was generated by running the following command:
+
[source,terminal]
----
$ oc config view --flatten
----
+
.Example externally generated `kubeconfig` file
+
[source,yaml]
----
apiVersion: v1
clusters:
- cluster:
    server: https://api.ci-ln-k0gim2b-76ef8.aws-2.ci.openshift.org:6443
  name: ci-ln-k0gim2b-76ef8
contexts:
- context:
    cluster: ci-ln-k0gim2b-76ef8
    user:
  name:
current-context:
kind: Config
preferences: {}
users:

- name:
  user:
    client-certificate-data: <removed>
    client-key-data:  <removed>
----
//does the microshift version have users listed in the file?
+
[NOTE]
====
The `certificate-authority-data` section is not present in externally generated `kubeconfig` files.
====

. Copy the externally generated certificates you want to add to the {microshift-short} host, ensuring that the files are accessible only by the trust root of the operating system.

> . Configure additional CAs in the trust root of the operating system. For example, in the RHEL Client truststore on the client system.

Reviewer 2: [link for configuring CAs in rhel trust store](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/securing_networks/using-shared-system-certificates_securing-networks):

+
[source,terminal]
----
$ export KUBECONFIG=/var/lib/microshift/resources/kubeadmin/<custom_cert_fqdn>/kubeconfig
----

. Check that the certificates are applied by using the following command:
+
[source,terminal]
----
$ oc --certificate-authority ~/certs/ca.ca get node
----
+
.Example output
[source,terminal]
----
oc get node
NAME                             STATUS   ROLES                         AGE   VERSION
dhcp-1-235-195.arm.example.com   Ready    control-plane,master,worker   76m   v1.29.2
----

. Check that the `openshift-service-ca` is running by entering the following command:
+
[source,terminal]
----
$ oc --insecure-skip-tls-verify get pod -A
----
+
.Example output
[source,terminal]
----
oc get pod -A
NAMESPACE                              NAME                                       READY   STATUS    RESTARTS      AGE
kube-system                            csi-snapshot-controller-5bdbd484f4-c6prb   1/1     Running   0             22m
kube-system                            csi-snapshot-webhook-957f6d687-vb7fl       1/1     Running   0             22m
openshift-dns                          dns-default-gbkx8                          2/2     Running   0             22m

> openshift-dns                          node-resolver-lqd7d                        1/1     Running   0             22m

Reviewer 2: I dont understand where  `openshift-service-ca` comes from, why do we need to have this section? Microshift doesnt use operators to manage its CAs, its being handled internally by Microshift  binary.

Author: This came from the Polarion. I didn't know where the line was between our testing v. user needed to do. We can remove.

Author: Looks like it was there because using the `oc get node` command gives the error `Unable to connect to the server: tls: failed to verify certificate: x509: certificate signed by unknown authority` is expected. Using `--insecure-skip-tls-verify` resolves the error.

openshift-operator-lifecycle-manager   olm-operator-b75ccb977-dh4wx               1/1     Running   0             22m
openshift-ovn-kubernetes               ovnkube-master-gr8xc                       4/4     Running   1 (22m ago)   22m
openshift-ovn-kubernetes               ovnkube-node-r9gth                         1/1     Running   1 (22m ago)   22m
openshift-service-ca                   service-ca-85fc4d5bb7-5g9vp                1/1     Running   0             22m
openshift-storage                      topolvm-controller-b94d5885b-7dmms         5/5     Running   0             22m
openshift-storage                      topolvm-node-x5hgs                         4/4     Running   0             22m
----
+
--
[NOTE]
====
If you use the `oc get node` command, the error `Unable to connect to the server: tls: failed to verify certificate: x509: certificate signed by unknown authority` is expected. Using `--insecure-skip-tls-verify` resolves the error.
====
--

. Add the new CA file to $KUBECONFIG by running the following command:
+
[source,terminal]
----
cat /tmp/caCert.pem /tmp/certificate-authority-data-origin.crt > /tmp/certificate-authority-data-new.crt

> sed -i "s/certificate-authority-data: .*/certificate-authority-data: $(base64 -w0 /tmp/certificate-authority-data-new.crt)/" $KUBECONFIG

Reviewer 2: this command will not  work for the generated kubeconfig  since they are missing `certificate-authority-data` so the  sed replace will not work.
its cleaner to use the oc command:
`oc config set clusters.microshift.certificate-authority /tmp/certificate-authority-data-new.crt`

----

. Confirm that the new CA has replaced the previous one by using the following steps:
.. After customizing the new CA, confirm that original certificate authority does not work by running the following command:
+
[source,termina]
----

> $ curl --cacert /tmp/certificate-authority-data-origin.crt https://${fqdn_name}:6443/healthz

Reviewer 2: in the previous step we changed the generated kubeconfig file content  to include the custom CA,
so it has not effect on the verification `curl` commands.
